table/merging_iterator.cc

```diff
@@ -1301,12 +1421,200 @@ void MergingIterator::InitMaxHeap() {
   }
 }
 
-// Repeatedly check and remove heap top key if it is not a point key
-// that is not covered by range tombstones. SeekImpl() is called to seek to end
-// of a range tombstone if the heap top is a point key covered by some range
-// tombstone from a newer sorted run. If the covering tombstone is from current
-// key's level, then the current child iterator is simply advanced to its next
-// key without reseeking.
+// Assume there is a next visible key.
+// Pre-condition:
+// - Invariants (4)-(10) hold
+// - There is some k where k <= children_[i].iter.key() <= LevelNextVisible(i,
+// k) for all levels i
+//
+// Define NextVisible(k) to be the first key >= k from among children_ that
+// is not covered by any range tombstone.
+// Post-condition:
+// - Invariants (2)-(10) hold
+// - minHeap_->top()->key() == NextVisible(k)
+//
+// Loop invariants:
+// - Invariant (4)-(10)
+// - (*): k <= children_[i].iter.key() <= LevelNextVisible(i, k)
+//
+// Progress: minHeap_.top()->key() is non-decreasing and strictly increases in
+// a finite number of iterations.
+// TODO: it is possible to call SeekImpl(k2) after SeekImpl(k1) with
+//  k2 < k1 in the same FindNextVisibleKey(). For example, l1 has a range
+//  tombstone [2,3) and l2 has a range tombstone [1, 4). Point key 1 from l5
+//  triggers SeekImpl(4 /* target */, 5). Then point key 2 from l3 triggers
+//  SeekImpl(3 /* target */, 3).
+//  Ideally we should only move iterators forward in SeekImpl(), and the
+//  progress condition can be made simpler: iterator only moves forward.
```

Comment:
Should the calls to `SeekImpl()` start at the range tombstone source level? That is, `SeekImpl(4, 2)` followed by `SeekImpl(3, 1)`. The latter shouldn't move backwards since its seek target will cascade to 4 for L2 and below.